### PR TITLE
Make Format Properties Lazy for dx12 backend

### PR DESCRIPTION
**Pull request by Veykril**
Progresses #2579

### CODE REVIEW

File: src/backend/dx12/src/device.rs [hunk 1]
```diff
@@ -2044,7 +2041,7 @@ impl d::Device<B> for Device {
         let block_dim = format_desc.dim;
         let extent = kind.extent();
 
-        let format_properties = &self.format_properties[format as usize];
+        let format_properties = &self.format_properties.get(format as usize);
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: nit: the reference is no longer needed here

File: src/backend/dx12/src/device.rs [hunk 2]
```diff
@@ -2215,7 +2212,7 @@ impl d::Device<B> for Device {
         //TODO: the clear_Xv is incomplete. We should support clearing images created without XXX_ATTACHMENT usage.
         // for this, we need to check the format and force the `RENDER_TARGET` flag behind the user's back
         // if the format supports being rendered into, allowing us to create clear_Xv
-        let format_properties = &self.format_properties[image_unbound.format as usize];
+        let format_properties = &self.format_properties.get(image_unbound.format as usize);
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: and here

File: src/backend/dx12/src/lib.rs [hunk 3]
```diff
@@ -1172,3 +1090,116 @@ fn validate_line_width(width: f32) {
...
+    fn drop(&mut self) {
+        unsafe {
+            self.1.destroy();
+        }
+    }
+}
+
+impl FormatProperties {
+    fn new(device: native::Device) -> Self {
+        let mut arr: [Mutex<Option<f::Properties>>; f::NUM_FORMATS] =
```
Comment (kvark) [inline src/backend/dx12/src/lib.rs]: let's just make it a `Box<[]>`, given that we can construct a `Vec` and convert it into that with https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.into_boxed_slice

File: src/backend/dx12/src/lib.rs [hunk 4]
```diff
@@ -1172,3 +1090,116 @@ fn validate_line_width(width: f32) {
...
+        for entry in arr.iter_mut().skip(1) {
+            unsafe { ptr::write(entry, Mutex::new(None)) };
+        }
+        FormatProperties(arr, device)
+    }
+
+    fn get(&self, idx: usize) -> f::Properties {
+        let mut guard = self.0[idx].lock().unwrap();
+        if let Some(props) = *guard {
+            props
```
Comment (kvark) [inline src/backend/dx12/src/lib.rs]: let's just `return` here to avoid indenting the rest of the code

File: src/hal/src/device.rs [hunk 5]
```diff
@@ -265,10 +265,16 @@ pub trait Device<B: Backend>: Any + Send + Sync {
     unsafe fn destroy_pipeline_layout(&self, layout: B::PipelineLayout);
 
     /// Create a pipeline cache object.
-    unsafe fn create_pipeline_cache(&self, data: Option<&[u8]>) -> Result<B::PipelineCache, OutOfMemory>;
+    unsafe fn create_pipeline_cache(
```
Comment (kvark) [inline src/hal/src/device.rs]: note: if you are applying rustfmt changes, please make it a separate commit independent of the functional changes

Comment (Veykril) [reply src/hal/src/device.rs]: Reverted them, didnt notice this when I ran it

**Comment by kvark** [comment]
bors retry
